[libc][math][c23] Add hypotf16 function

meltq opened the pull request:
Implement hypot for Float16 along with tests.

meltq committed:
Add hypotf16() function

meltq committed:
Formatting changes

meltq commented:
WIP. `test_subnormal_range` in the unit tests takes forever to run on my system for some reason, there seems to be some problem with `sNaN` being converted to a quiet NaN after `FPBits(x).abs()`.

meltq committed:
Temp removed exhaustive

meltq committed:
Formatting changes

meltq committed:
Add normal and subnormal tests

meltq committed:
Add 1 to round up STEP for subnormal

meltq committed:
Added hard to round, formatting

meltq committed:
Formatting changes

meltq commented:
@overmighty @lntue requesting review.

lntue commented on libc/test/src/math/hypotf16_test.cpp:
```diff
@@ -0,0 +1,27 @@
+//===----------------------------------------------------------------------===//
+
+#include "HypotTest.h"
+#include "hypotf16_hard_to_round.h"
```
For `hypotf16`, we can actually test it exhaustively, so you don't need to add hard-to-round test.  `hypotf16_hard_to_round.h` could be removed.

meltq replied:
Isn't it still useful to have hard to round cases for it to be run during the normal `check-libc` test? Running exhaustive tests on my system takes around 40 minutes with all the types of rounding, so hard_to_round seems like a good intermediary set of tests someone could run while changing the implementation before running exhaustive tests.

lntue commented on libc/test/src/math/exhaustive/hypotf16_test.cpp:
```diff
@@ -0,0 +1,58 @@
+        float16 y = FPBits(ybits).get_val();
+        bool correct = TEST_FP_EQ(LIBC_NAMESPACE::fputil::hypot(x, y),
+                                  LIBC_NAMESPACE::hypotf16(x, y));
+        // Using MPFR will be much slower.
```
How long did it take for you to run the exhaustive tests? With and without mpfr?

meltq replied:
Without MPFR it took 46 minutes. With MPFR, it took 1 hour 32 minutes. This is on an AMD Ryzen 7 7730U processor.

meltq committed:
Added increment, cleanup

meltq commented:
Added increment which parallelizes the exhaustive test and reduces time taken from 46 minutes to 7 minutes. Added other suggested changes too.

lntue commented:
I've got the following warnings when building the tests.  Can you add explicit casts for them?
```
[12/19] Building CXX object projects/libc/test/src/math/exhaustive/C...t.src.math.exhaustive.hypotf16_test.__unit__.dir/hypotf16_test.cpp.o
In file included from /usr/local/google/home/lntue/experiment/llvm-project/libc/test/src/math/exhaustive/hypotf16_test.cpp:11:
/usr/local/google/home/lntue/experiment/llvm-project/libc/src/__support/FPUtil/Hypot.h:210:18: warning: implicit conversion loses integer precision: 'int' to 'StorageType' (aka 'unsigned short') [-Wimplicit-int-conversion]
    r = (r << 1) + ((tail_bits & current_bit) ? 1 : 0);
      ~ ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/lntue/experiment/llvm-project/libc/test/src/math/exhaustive/hypotf16_test.cpp:38:59: note: in instantiation of function template specialization '__llvm_libc_21_0_0_git::fputil::hypot<_Float16, 0>' requested here
        bool correct = TEST_FP_EQ(LIBC_NAMESPACE::fputil::hypot(x, y),
                                                          ^
In file included from /usr/local/google/home/lntue/experiment/llvm-project/libc/test/src/math/exhaustive/hypotf16_test.cpp:11:
/usr/local/google/home/lntue/experiment/llvm-project/libc/src/__support/FPUtil/Hypot.h:211:36: warning: implicit conversion loses integer precision: 'int' to 'StorageType' (aka 'unsigned short') [-Wimplicit-int-conversion]
    StorageType tmp = (y_new << 1) + current_bit; // 2*y_new(n - 1) + 2^(-n)
                ~~~   ~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/local/google/home/lntue/experiment/llvm-project/libc/src/__support/FPUtil/Hypot.h:33:15: warning: implicit conversion loses integer precision: 'int' to 'unsigned short' [-Wimplicit-int-conversion]
  return T(1) << shift_length;
  ~~~~~~ ~~~~~^~~~~~~~~~~~~~~
/usr/local/google/home/lntue/experiment/llvm-project/libc/src/__support/FPUtil/Hypot.h:158:29: note: in instantiation of function template specialization '__llvm_libc_21_0_0_git::fputil::internal::find_leading_one<unsigned short>' requested here
    leading_one = internal::find_leading_one(a_mant, y_mant_width);
                            ^
/usr/local/google/home/lntue/experiment/llvm-project/libc/test/src/math/exhaustive/hypotf16_test.cpp:38:59: note: in instantiation of function template specialization '__llvm_libc_21_0_0_git::fputil::hypot<_Float16, 0>' requested here
        bool correct = TEST_FP_EQ(LIBC_NAMESPACE::fputil::hypot(x, y),
```

lntue commented on libc/test/src/math/exhaustive/hypotf16_test.cpp:
```diff
@@ -0,0 +1,58 @@
+namespace mpfr = LIBC_NAMESPACE::testing::mpfr;
+
+// Range of both inputs: [0, inf]
+static constexpr uint16_t START = 0x0000U;
```
The tests are very fast now, can you also test for negative inputs?

meltq committed:
Added perf test

meltq committed:
formatting changes

meltq committed:
Replace fputil::cast with static_cast

meltq committed:
Removed hard to round cases

meltq committed:
Added negative exhaustive test

lntue commented on libc/src/math/generic/hypotf16.cpp:
```diff
@@ -51,8 +51,8 @@ LLVM_LIBC_FUNCTION(float16, hypotf16, (float16 x, float16 y)) {
 
-  float af = fputil::cast<float>(a_bits.get_val());
-  float bf = fputil::cast<float>(b_bits.get_val());
+  float af = static_cast<float>(a_bits.get_val());
```
don't replace `fputil::cast` with `static_cast`.  It is there because there are several targets with not correctly rounded `cast` in the compiler runtime that backing up these `static_cast`.  Instead we should add specializations / optimizations for `fputil::cast` for known-good targets in the followup work.

lntue replied:
You can add comments on the top of the function, mentioning that for targets in which the casts between float16 and float have to be fully emulated, `fputil::hypot<float16>` would be faster.

meltq committed:
Add explicit casts in Hypot

meltq committed:
Add back fputil::cast

meltq committed:
Add comments regarding potential slow down

meltq commented:
Added negative exhaustive tests, removed `hypotf16_hard_to_round` since exhaustive tests would finish executing much faster if `hypotf16` used an optimized version of `fputil::cast` that didn't emulate (14 minutes -> 1 min 45 s on my system). Added a comment suggesting `fputil::hypot<float16>` for cases where cast needs to be emulated, created and linked an issue regarding the optimization of `fputil::cast` https://github.com/llvm/llvm-project/issues/133517

meltq committed:
Formatting changes

meltq committed:
Formatting changes

lntue commented on libc/src/__support/FPUtil/Hypot.h:
```diff
@@ -207,8 +207,9 @@ LIBC_INLINE T hypot(T x, T y) {
        current_bit >>= 1) {
-    r = (r << 1) + ((tail_bits & current_bit) ? 1 : 0);
-    StorageType tmp = (y_new << 1) + current_bit; // 2*y_new(n - 1) + 2^(-n)
+    r = (StorageType)(r << 1) + ((tail_bits & current_bit) ? 1 : 0);
+    StorageType tmp =
+        (StorageType)(y_new << 1) + current_bit; // 2*y_new(n - 1) + 2^(-n)
```
nit: use C++ cast instead of C cast, either `static_cast<>` or `StorageType()`

lntue commented on libc/src/__support/FPUtil/Hypot.h:
```diff
@@ -30,7 +30,7 @@ LIBC_INLINE T find_leading_one(T mant, int &shift_length) {
     shift_length = (sizeof(mant) * 8) - 1 - cpp::countl_zero(mant);
   }
-  return T(1) << shift_length;
+  return (T)(T(1) << shift_length);
```
nit: use C++ cast

meltq committed:
Use static_cast

meltq committed:
Formatting changes

meltq commented:
Changed to using `static_cast` for explicit casts.
